status-whisper-usage-spec.md

```diff
@@ -245,6 +246,29 @@ Public and group messages are encrypted using symmetric encryption and the key i
 
 One-to-one messages are encrypted using asymmetric encryption.
 
+## Message confirmations
```

Comment:
This API in general is quite confusion to me. Is there any history of how this came to be? What's the rationale? How does it interact with MVDS? Doesn't it lead to a lot of extra traffic? Do we definitely want to keep it? Do we dependent on it?

Reply:
A lot of questions :D 

> Is there any history of how this came to be? What's the rationale?

I don't recall anything, I haven't participated in the implementation, I don't know why we have to packet codes.

> Why does it do ad hoc message sync when we have MVDS for that on top?

Reply:
I too find this quite a strange addition to the API. 
- I don't understand how this actually confirms anything. If used only for 1 hop, a confirmation does not give you any guarantee of arrival to the real recipients. 
- It would introduce a whole lot of traffic when applied on all nodes (for each hop again). (I read that it is only active for Mail servers ? )

So I guess it is now just used for mail servers so you are sure that your message gets in a mailserver database there, but only if you are directly connected with those, which in the current setup, I guess you are. Thisis the kind of functionality, that is really based too much on the network setup imo.

Also, some thought regarding privacy and security implications would be wise here.

Anyway, I understand that this is the situation as is, but for Waku/0 or Waku/1 we should definitely revise whether this is still needed. And not already handled by the layer up (as I believe it should be when a transport as Whisper is used).

Reply:
> I don't understand how this actually confirms anything. If used only for 1 hop, a confirmation does not give you any guarantee of arrival to the real recipients.

No, it only gives you a confirmation that a message has been received by at least one node. In our case, this node is mailserver.

> It would introduce a whole lot of traffic when applied on all nodes (for each hop again).

That's why it's not broadcast-based.

> Thisis the kind of functionality, that is really based too much on the network setup imo.

Generally yes. It assumes that you are directly connected to a mailserver and actually if you want to use mailserver, you have to be directly connected to one.

> Anyway, I understand that this is the situation as is, but for Waku/0 or Waku/1 we should definitely revise whether this is still needed. And not already handled by the layer up (as I believe it should be when a transport as Whisper is used).

Correct. Like I mentioned in some other comment, this solution was introduced rather quickly and without collecting proper thinking. It's definitely not something we should be really attached to. 

This is more general problem of storing offline messages. The granularity of confirmations might be also increased to improve the UX.

Reply:
i can try to explain, status chat notifies user when message was sent to the network. and user is able to retry only if message failed to be ack'ed by the network.

the rationale was that you never know if the message was actually sent to the network or not, socket api returns immediately when message was written to a socket buffer. so there was a ton of false positives and this solution was added.

mvds gives you confirmation that the message was actually delivered to the end recipient and it is a much better guarantee then this solution, but not quite the same. you can think of it like a first confirmation from telegram server, which is obviously a different model.

as for bw confirmation was made for whole batch of whisper envelopes, so this is like 32 bytes overhead per single batch. not broadcasted like it was noted by Adam. just 32 bytes directly written to a sender of the batch. pretty sure that amount is negligible in comparison with other whisper traffic

Reply:
btw i am not advocating for this solution to stay or smth like that and perfectly understand that it provides some guarantees only because of how status cluster work. just wanted to provide some intuition on how it came to be.